src/Model/DesignDocumentModel.hpp

```diff
@@ -0,0 +1,2450 @@
+  {
+    if (j.is_null())
+      return std::make_shared<T>();
+    else
+      return std::make_shared<T>(j.get<T>());
+  }
+};
+template<typename T>
+struct adl_serializer<std::optional<T>>
+{
+  static void to_json(json& j, const std::optional<T>& opt)
+  {
+    if (!opt)
+      j = nullptr;
+    else
+      j = *opt;
+  }
+
+  static std::optional<T> from_json(const json& j)
+  {
+    if (j.is_null())
+      return std::make_optional<T>();
+    else
+      return std::make_optional<T>(j.get<T>());
+  }
+};
+} // namespace nlohmann
+#endif
+
+namespace VGG
+{
+using nlohmann::json;
+
+#ifndef NLOHMANN_UNTYPED_VGG_HELPER
+#define NLOHMANN_UNTYPED_VGG_HELPER
+inline json get_untyped(const json& j, const char* property)
+{
+  if (j.find(property) != j.end())
+  {
+    return j.at(property).get<json>();
+  }
+  return json();
+}
+
+inline json get_untyped(const json& j, std::string property)
+{
+  return get_untyped(j, property.data());
+}
+#endif
+
+#ifndef NLOHMANN_OPTIONAL_VGG_HELPER
+#define NLOHMANN_OPTIONAL_VGG_HELPER
+template<typename T>
+inline std::shared_ptr<T> get_heap_optional(const json& j, const char* property)
+{
+  auto it = j.find(property);
+  if (it != j.end() && !it->is_null())
+  {
+    return j.at(property).get<std::shared_ptr<T>>();
+  }
+  return std::shared_ptr<T>();
+}
+
+template<typename T>
+inline std::shared_ptr<T> get_heap_optional(const json& j, std::string property)
+{
+  return get_heap_optional<T>(j, property.data());
+}
+template<typename T>
+inline std::optional<T> get_stack_optional(const json& j, const char* property)
+{
+  auto it = j.find(property);
+  if (it != j.end() && !it->is_null())
+  {
+    return j.at(property).get<std::optional<T>>();
+  }
+  return std::optional<T>();
+}
+
+template<typename T>
+inline std::optional<T> get_stack_optional(const json& j, std::string property)
+{
+  return get_stack_optional<T>(j, property.data());
+}
+#endif
+
+/**
+ * Defines a RGBA color value.
+ *
+ * The border color of the font.If it doesn't exist it means it doesn't need to be stroked.
+ *
+ * The fill color of the font.If it doesn't exist it means it doesn't need to be filled.
+ *
+ * The color of the shadow.
+ */
+struct Color
+{
+  double alpha;
+  double blue;
+  nlohmann::json color_class;
+  double green;
+  double red;
+};
+
+/**
+ * This property records the bounds information of the object before it is matrixed. For
+ * example, In adobe illustrator, the top left coordinate of an image is always 0, So the
+ * relationship of the bounds frame matrix is as follows: { (0, 0), (width, 0), (width,
+ * -height), (0, -height) } + matrix_transformation = frame.
+ *
+ * Defines an abstract rectangle.
+ *
+ * Frame information for the object. When the object is a group, its frame is the union of
+ * all objects in the group, whether or not the group contains a mask. This property stores
+ * information about the object after the matrix transformation.
+ */
+struct Rect
+{
+  nlohmann::json rect_class;
```

Comment:
Could we handle such const field properly?

Reply:
Yes, we will handle it carefully. 
Construct this struct from json need public field, so const filed is not possible. 
The field's value will be validate if changed.